# Stock Data Cleaning & Preliminary Analysis

Can we predict Google's stock price with machine learning?

## Loading in Data and Checking Dataframes

In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt
import glob


In [7]:
datasets = glob.glob("C:/Users/linwi/Documents/Stock_test/Datasets/*.csv")
datasets


['C:/Users/linwi/Documents/Stock_test/Datasets\\sp500_companies.csv',
 'C:/Users/linwi/Documents/Stock_test/Datasets\\sp500_index.csv',
 'C:/Users/linwi/Documents/Stock_test/Datasets\\sp500_stocks.csv']

In [8]:
dfs = []
for file in datasets:
    # Read the CSV file into a DataFrame
    df = pd.read_csv(file)

    # Append the DataFrame to the list
    dfs.append(df)

In [9]:
dfs #Checking if all the dataframes are in the list 

[    Exchange Symbol                  Shortname  \
 0        NMS   MSFT      Microsoft Corporation   
 1        NMS   AAPL                 Apple Inc.   
 2        NMS   NVDA         NVIDIA Corporation   
 3        NMS   AMZN           Amazon.com, Inc.   
 4        NMS   GOOG              Alphabet Inc.   
 ..       ...    ...                        ...   
 497      NYQ   GNRC     Generac Holdlings Inc.   
 498      NYQ    CMA      Comerica Incorporated   
 499      NYQ    VFC           V.F. Corporation   
 500      NYQ    WHR      Whirlpool Corporation   
 501      NMS   ZION  Zions Bancorporation N.A.   
 
                                        Longname                  Sector  \
 0                         Microsoft Corporation              Technology   
 1                                    Apple Inc.              Technology   
 2                            NVIDIA Corporation              Technology   
 3                              Amazon.com, Inc.       Consumer Cyclical   
 4    

In [10]:
sp500_df = dfs[1].copy() #Slicing out the S&P500 dataframe and making a copy.

In [11]:
sp500_df

Date   S&P500
0     2014-03-03  1845.73
1     2014-03-04  1873.91
2     2014-03-05  1873.81
3     2014-03-06  1877.03
4     2014-03-07  1878.04
...          ...      ...
2513  2024-02-26  5069.53
2514  2024-02-27  5078.18
2515  2024-02-28  5069.76
2516  2024-02-29  5096.27
2517  2024-03-01  5137.08

[2518 rows x 2 columns]

We can see that the sp500_df has two columns: `Date` and `S&P500`

In [12]:
company_df = dfs[0].copy() #Slicing out the company dataframe and making a copy.

In [13]:
company_df

Exchange Symbol                  Shortname  \
0        NMS   MSFT      Microsoft Corporation   
1        NMS   AAPL                 Apple Inc.   
2        NMS   NVDA         NVIDIA Corporation   
3        NMS   AMZN           Amazon.com, Inc.   
4        NMS   GOOG              Alphabet Inc.   
..       ...    ...                        ...   
497      NYQ   GNRC     Generac Holdlings Inc.   
498      NYQ    CMA      Comerica Incorporated   
499      NYQ    VFC           V.F. Corporation   
500      NYQ    WHR      Whirlpool Corporation   
501      NMS   ZION  Zions Bancorporation N.A.   

                                       Longname                  Sector  \
0                         Microsoft Corporation              Technology   
1                                    Apple Inc.              Technology   
2                            NVIDIA Corporation              Technology   
3                              Amazon.com, Inc.       Consumer Cyclical   
4                                 Alphabet Inc.  Communication Services   
..                                          ...                     ...   
497                       Generac Holdings Inc.             Industrials   
498                       Comerica Incorporated      Financial Services   
499                            V.F. Corporation       Consumer Cyclical   
500                       Whirlpool Corporation       Consumer Cyclical   
501  Zions Bancorporation, National Association      Financial Services   

                               Industry  Currentprice      Marketcap  \
0             Software - Infrastructure        415.50  3087347744768   
1                  Consumer Electronics        179.66  2774291709952   
2                        Semiconductors        822.79  2056974893056   
3                       Internet Retail        178.22  1851242446848   
4        Internet Content & Information        138.08  1710397063168   
..                                  ...           ...            ...   
497      Specialty Industrial Machinery        114.18     6881548800   
498                    Banks - Regional         49.59     6540822016   
499               Apparel Manufacturing         16.24     6314371584   
500  Furnishings, Fixtures & Appliances        106.81     5836504064   
501                    Banks - Regional         38.77     5743969792   

           Ebitda  Revenuegrowth            City State        Country  \
0    1.184270e+11          0.176         Redmond    WA  United States   
1    1.301090e+11          0.021       Cupertino    CA  United States   
2    3.448000e+10          2.653     Santa Clara    CA  United States   
3    8.551500e+10          0.139         Seattle    WA  United States   
4    1.001720e+11          0.135   Mountain View    CA  United States   
..            ...            ...             ...   ...            ...   
497  5.532610e+08          0.014        Waukesha    WI  United States   
498           NaN         -0.220          Dallas    TX  United States   
499  1.068373e+09         -0.162          Denver    CO  United States   
500  1.477000e+09          0.034   Benton Harbor    MI  United States   
501           NaN         -0.119  Salt Lake City    UT  United States   

     Fulltimeemployees                                Longbusinesssummary  \
0             221000.0  Microsoft Corporation develops and supports so...   
1             161000.0  Apple Inc. designs, manufactures, and markets ...   
2              29600.0  NVIDIA Corporation provides graphics, and comp...   
3            1525000.0  Amazon.com, Inc. engages in the retail sale of...   
4             182502.0  Alphabet Inc. offers various products and plat...   
..                 ...                                                ...   
497             8315.0  Generac Holdings Inc. designs, manufactures, a...   
498             7701.0  Comerica Incorporated, through its subsidiarie...   
499            19800.0  V.F. Corporation, together with its subsidi

The company_df contains fundamental information of the companies in the S&P500. Its categories are: `Exchange`,`Symbol`,`Shortname`,`Longname`, `Sector`, `Industry`,`Currentprice`, `Marketcap`, `Ebitda`, `Revenuegrowth`, `City`, `State`, `Country`, `Fulltimeemployees`, `Longbusinesssummary` and `Weight`.

In [14]:
stocks_df = dfs[2].copy()

In [15]:
stocks_df

Date Symbol   Adj Close       Close        High         Low  \
0        2010-01-04    MMM   53.295380   83.019997   83.449997   82.669998   
1        2010-01-05    MMM   52.961567   82.500000   83.230003   81.699997   
2        2010-01-06    MMM   53.712662   83.669998   84.599998   83.510002   
3        2010-01-07    MMM   53.751194   83.730003   83.760002   82.120003   
4        2010-01-08    MMM   54.129951   84.320000   84.320000   83.300003   
...             ...    ...         ...         ...         ...         ...   
1792687  2024-02-26    ZTS  197.039993  197.039993  197.990005  195.460007   
1792688  2024-02-27    ZTS  199.000000  199.000000  199.250000  194.039993   
1792689  2024-02-28    ZTS  199.940002  199.940002  200.500000  197.100006   
1792690  2024-02-29    ZTS  198.330002  198.330002  200.529999  198.190002   
1792691  2024-03-01    ZTS  190.820007  190.820007  196.229996  188.490005   

               Open     Volume  
0         83.089996  3043700.0  
1         82.800003  2847000.0  
2         83.879997  5268500.0  
3         83.320000  4470100.0  
4         83.690002  3405800.0  
...             ...        ...  
1792687  196.899994  2098500.0  
1792688  197.649994  1982500.0  
1792689  199.300003  2003300.0  
1792690  200.500000  2849200.0  
1792691  196.000000  3724300.0  

[1792692 rows x 8 columns]

The stocks_df has `Date`, `Symbol`, `Adj Close`, `Close`, `High`, `Low` , `Open`, and `Volume` 

**Checking the datatypes in each dataframe.**

In [16]:
stocks_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1792692 entries, 0 to 1792691
Data columns (total 8 columns):
 #   Column     Dtype  
---  ------     -----  
 0   Date       object 
 1   Symbol     object 
 2   Adj Close  float64
 3   Close      float64
 4   High       float64
 5   Low        float64
 6   Open       float64
 7   Volume     float64
dtypes: float64(6), object(2)
memory usage: 109.4+ MB


In [17]:
company_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 502 entries, 0 to 501
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Exchange             502 non-null    object 
 1   Symbol               502 non-null    object 
 2   Shortname            502 non-null    object 
 3   Longname             502 non-null    object 
 4   Sector               502 non-null    object 
 5   Industry             502 non-null    object 
 6   Currentprice         502 non-null    float64
 7   Marketcap            502 non-null    int64  
 8   Ebitda               473 non-null    float64
 9   Revenuegrowth        501 non-null    float64
 10  City                 502 non-null    object 
 11  State                482 non-null    object 
 12  Country              502 non-null    object 
 13  Fulltimeemployees    470 non-null    float64
 14  Longbusinesssummary  502 non-null    object 
 15  Weight               502 non-null    flo

In [18]:
sp500_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2518 entries, 0 to 2517
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Date    2518 non-null   object 
 1   S&P500  2518 non-null   float64
dtypes: float64(1), object(1)
memory usage: 39.5+ KB


## Data Cleaning 

In [46]:
print(stocks_df.isna().sum())

Date             0
Symbol           0
Adj Close    79913
Close        79913
High         79913
Low          79913
Open         79913
Volume       79913
dtype: int64


In [47]:
print(sp500_df.isna().sum())

Date      0
S&P500    0
dtype: int64


In [48]:
print(company_df.isna().sum())

Exchange                0
Symbol                  0
Shortname               0
Longname                0
Sector                  0
Industry                0
Currentprice            0
Marketcap               0
Ebitda                 29
Revenuegrowth           1
City                    0
State                  20
Country                 0
Fulltimeemployees      32
Longbusinesssummary     0
Weight                  0
dtype: int64


We can see that there are 79,913 null values in the stocks_df.
There are no null values in the sp500_df.
There are some null values for the company_df.

Let's first look at the null values for the stocks_df.

In [41]:
stocks_null = stocks_df.isna().any(axis=1)
stocks_null_rows = stocks_df[stocks_null]
print(stocks_null_rows)

               Date Symbol  Adj Close  Close  High  Low  Open  Volume
10692    2010-01-04   ABBV        NaN    NaN   NaN  NaN   NaN     NaN
10693    2010-01-05   ABBV        NaN    NaN   NaN  NaN   NaN     NaN
10694    2010-01-06   ABBV        NaN    NaN   NaN  NaN   NaN     NaN
10695    2010-01-07   ABBV        NaN    NaN   NaN  NaN   NaN     NaN
10696    2010-01-08   ABBV        NaN    NaN   NaN  NaN   NaN     NaN
...             ...    ...        ...    ...   ...  ...   ...     ...
1789898  2013-01-25    ZTS        NaN    NaN   NaN  NaN   NaN     NaN
1789899  2013-01-28    ZTS        NaN    NaN   NaN  NaN   NaN     NaN
1789900  2013-01-29    ZTS        NaN    NaN   NaN  NaN   NaN     NaN
1789901  2013-01-30    ZTS        NaN    NaN   NaN  NaN   NaN     NaN
1789902  2013-01-31    ZTS        NaN    NaN   NaN  NaN   NaN     NaN

[79913 rows x 8 columns]


The null values look like they are the same across each row and column.

In [52]:
company_null = company_df.isna().any(axis=1)
company_null_rows = company_df[company_null]
print(company_null_rows)

    Exchange Symbol                        Shortname  \
11       NYQ      V                        Visa Inc.   
12       NYQ    JPM            JP Morgan Chase & Co.   
14       NYQ    UNH  UnitedHealth Group Incorporated   
16       NYQ    XOM          Exxon Mobil Corporation   
17       NYQ    JNJ                Johnson & Johnson   
..       ...    ...                              ...   
463      NYQ   ALLE                     Allegion plc   
471      NYQ   BBWI          Bath & Body Works, Inc.   
494      NMS   XRAY             DENTSPLY SIRONA Inc.   
498      NYQ    CMA            Comerica Incorporated   
501      NMS   ZION        Zions Bancorporation N.A.   

                                       Longname              Sector  \
11                                    Visa Inc.  Financial Services   
12                         JPMorgan Chase & Co.  Financial Services   
14              UnitedHealth Group Incorporated          Healthcare   
16                      Exxon Mobil Corpora

## Google Analysis and Modeling

In [28]:
#Selecting only the rows with the Symbol "GOOG"
goog_df = stocks_df[stocks_df["Symbol"]== 'GOOG']
type(goog_df)

pandas.core.frame.DataFrame

In [50]:
goog_df.isna().sum() #checking if the goog dataframe has any nullvalues.

Date         0
Symbol       0
Adj Close    0
Close        0
High         0
Low          0
Open         0
Volume       0
dtype: int64

We will narrow down our analysis and focus on the 'GOOG' stock. 

In [29]:
goog_df

Date Symbol   Adj Close       Close        High         Low  \
71280  2010-01-04   GOOG   15.610239   15.610239   15.678981   15.547723   
71281  2010-01-05   GOOG   15.541497   15.541497   15.637387   15.480475   
71282  2010-01-06   GOOG   15.149715   15.149715   15.588072   15.102393   
71283  2010-01-07   GOOG   14.797037   14.797037   15.193053   14.760922   
71284  2010-01-08   GOOG   14.994298   14.994298   15.024933   14.672753   
...           ...    ...         ...         ...         ...         ...   
74839  2024-02-26   GOOG  138.750000  138.750000  143.839996  138.740005   
74840  2024-02-27   GOOG  140.100006  140.100006  140.490005  138.500000   
74841  2024-02-28   GOOG  137.429993  137.429993  139.279999  136.639999   
74842  2024-02-29   GOOG  139.779999  139.779999  139.949997  137.570007   
74843  2024-03-01   GOOG  138.080002  138.080002  140.000000  137.975006   

             Open       Volume  
71280   15.615220   78541293.0  
71281   15.620949  120638494.0  
71282   15.588072  159744526.0  
71283   15.178109  257533695.0  
71284   14.744733  189680313.0  
...           ...          ...  
74839  143.449997   33513000.0  
74840  139.410004   22364000.0  
74841  139.100006   30628700.0  
74842  138.350006   35485000.0  
74843  139.610001   28522000.0  

[3564 rows x 8 columns]